frontend_vm/src/ast.rs

```diff
@@ -0,0 +1,205 @@
+//! Common AST for the frontend architecture inputs.
+
+use std::fmt::{self, Display};
+
+#[derive(Clone)]
+pub enum Statement {
+    Label(String),
+    Directive(String, Vec<Argument>),
+    Instruction(String, Vec<Argument>),
+}
+
+#[derive(Clone)]
+pub enum Argument {
+    Register(Register),
+    RegOffset(Register, Expression),
+    StringLiteral(Vec<u8>),
+    Expression(Expression),
+}
+
+impl Argument {
+    pub fn post_visit_expressions_mut(&mut self, f: &mut impl FnMut(&mut Expression)) {
+        match self {
+            Argument::Register(_) | Argument::StringLiteral(_) => (),
+            Argument::RegOffset(_, expr) | Argument::Expression(expr) => {
+                expr.post_visit_mut(f);
+            }
+        }
+    }
+
+    pub fn post_visit_expressions<'a>(&'a self, f: &mut impl FnMut(&'a Expression)) {
+        match self {
+            Argument::Register(_) | Argument::StringLiteral(_) => (),
+            Argument::RegOffset(_, expr) | Argument::Expression(expr) => {
+                expr.post_visit(f);
+            }
+        }
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub struct Register(pub u8);
+
+impl Register {
+    pub fn is_zero(&self) -> bool {
```

Comment:
We can also do that later, but this function is mainly useful on riscv, since there is no special "zero register" in e.g. the intel architecture. We have to see how to do the register mapping anyway.

Reply:
Maybe it would be best to make the register type a generic parameter or something.